src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringConcat.cs

```diff
@@ -299,33 +334,208 @@ private BoundExpression RewriteStringConcatenationOneExpr(BoundExpression lowere
 
         private BoundExpression RewriteStringConcatenationTwoExprs(SyntaxNode syntax, BoundExpression loweredLeft, BoundExpression loweredRight)
         {
-            Debug.Assert(loweredLeft.HasAnyErrors || loweredLeft.Type is { } && loweredLeft.Type.IsStringType());
-            Debug.Assert(loweredRight.HasAnyErrors || loweredRight.Type is { } && loweredRight.Type.IsStringType());
+            var leftType = loweredLeft.Type;
+            var rightType = loweredRight.Type;
+
+            Debug.Assert(loweredLeft.HasAnyErrors || leftType is not null && (leftType.IsStringType() || leftType.IsCharType()));
+            Debug.Assert(loweredRight.HasAnyErrors || rightType is not null && (rightType.IsStringType() || rightType.IsCharType()));
+
+            var leftIsChar = leftType?.IsCharType() == true;
+            var rightIsChar = rightType?.IsCharType() == true;
+
+            if ((leftIsChar || rightIsChar) &&
+                TryGetWellKnownTypeMember(syntax, WellKnownMember.System_String__Concat_ReadOnlySpanReadOnlySpan, out MethodSymbol stringConcatSpans, isOptional: true) &&
+                TryGetWellKnownTypeMember(syntax, WellKnownMember.System_String__op_Implicit_ToReadOnlySpanOfChar, out MethodSymbol stringImplicitConversionToReadOnlySpan, isOptional: true) &&
+                TryGetWellKnownTypeMember(syntax, WellKnownMember.System_ReadOnlySpan_T__ctor_Reference, out MethodSymbol readOnlySpanCtorRefParamGeneric, isOptional: true))
+            {
+                // One of args is a char. We can use span-based concatenation, which takes a bit more IL, but avoids allocation of a string from ToString call.
+                // And since implicit conversion from string to span is a JIT intrinsic, the resulting code ends up being faster than the one generated from the "naive" case with ToString call
+                var readOnlySpanOfChar = readOnlySpanCtorRefParamGeneric.ContainingType.Construct(_compilation.GetSpecialType(SpecialType.System_Char));
```

Comment:
>_compilation.GetSpecialType(SpecialType.System_Char) [](http://example.com/codeflow?start=98&length=52)

```leftIsChar ? leftType : rightType```?

This is likely to be faster, and there is no question why we do not collect use-site info for the type.